src/main/kotlin/org/elm/ide/inspections/ElmUnusedImportInspection.kt

```diff
@@ -0,0 +1,78 @@
+package org.elm.ide.inspections
+
+import com.intellij.codeInspection.LocalInspectionTool
+import com.intellij.codeInspection.LocalInspectionToolSession
+import com.intellij.codeInspection.ProblemHighlightType
+import com.intellij.codeInspection.ProblemsHolder
+import com.intellij.openapi.util.Key
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiElementVisitor
+import org.elm.lang.core.psi.ElmExposedItemTag
+import org.elm.lang.core.psi.ElmFile
+import org.elm.lang.core.psi.elements.ElmImportClause
+import org.elm.lang.core.resolve.ElmReferenceElement
+import org.elm.lang.core.resolve.scope.ModuleScope
+
+/**
+ * Find unused imports
+ */
+class ElmUnusedImportInspection : LocalInspectionTool() {
+
+    private val visitorKey = Key.create<ImportVisitor>("")
+
+    override fun buildVisitor(holder: ProblemsHolder, isOnTheFly: Boolean, session: LocalInspectionToolSession): PsiElementVisitor {
+        val file = session.file as? ElmFile
+                ?: return super.buildVisitor(holder, isOnTheFly, session)
+        val visitor = ImportVisitor(ModuleScope(file).getImportDecls())
+        session.putUserData(visitorKey, visitor)
+        return visitor
+    }
+
+    override fun inspectionFinished(session: LocalInspectionToolSession, problemsHolder: ProblemsHolder) {
+        val visitor = session.getUserData(visitorKey) ?: return
+        for (unusedImport in visitor.unusedImports) {
+            markAsUnused(problemsHolder, unusedImport)
+        }
+    }
+
+    private fun markAsUnused(holder: ProblemsHolder, importClause: ElmImportClause) {
+        holder.registerProblem(
+                importClause,
+                "Unused import",
+                ProblemHighlightType.LIKE_UNUSED_SYMBOL
+        )
+    }
+
+    private fun markAsUnused(holder: ProblemsHolder, exposedItem: ElmExposedItemTag) {
+        holder.registerProblem(
+                exposedItem,
+                "'${exposedItem.text}' is unused",
+                ProblemHighlightType.LIKE_UNUSED_SYMBOL
+        )
+    }
+}
+
+class ImportVisitor(initialImports: List<ElmImportClause>) : PsiElementVisitor() {
+
+    // IMPORTANT! IntelliJ's LocalInspectionTool requires that visitor implementations be thread-safe.
+    private val imports = initialImports.toMutableList()
```

Comment:
Now that I think about it, maybe I should use a ConcurrentHashMap indexed by the module name string.